### PR TITLE
Move away from pubsub client

### PR DESCRIPTION
**Pull request by eplanet**

### CODE REVIEW

**Comment by antonipp** [comment]
LGTM for testing purposes but if we want to have this in production, we will have to add error handling and exponential backoff 👍

File: luigi/pubsub_history.py [hunk 1]
```diff
@@ -34,14 +36,17 @@ class PubsubHistory(object):
     """ A History class that sends messages to a Pub/Sub topic
     """
     def __init__(self):
-        self.publisher = pubsub.PublisherClient()
         self.config = configuration.get_config()
+        self.credentials = service_account.Credentials.from_service_account_file(os.environ['GOOGLE_APPLICATION_CREDENTIALS'], scopes=['https://www.googleapis.com/auth/pubsub', 'https://www.googleapis.com/auth/cloud-platform'])
```
Comment (antonipp) [inline luigi/pubsub_history.py]: This probably doesn't need to be a class attribute, it can just be a local variable

Comment (eplanet) [reply luigi/pubsub_history.py]: I think it's good to refresh it within the backoff, but I'd rather avoid that type of call for each query.

File: luigi/pubsub_history.py [hunk 2]
```diff
@@ -14,34 +14,95 @@
...
+    the function is passed as the first argument and decorate returns the decorated function, as expected.
+    If the decorator is called with one or more optional arguments like this:
+    @decorator(optional_argument1='some value')
+    def function ....
+    Then decorator is called with the function argument with value None, so a function that decorates is returned, as expected.
+    see https://stackoverflow.com/questions/3888158/making-decorators-with-optional-arguments for more details
+    :param timeout: maximum duration of the backoff, in seconds
+    """
+
+    def _decorate(func):
```
Comment (antonipp) [inline luigi/pubsub_history.py]: This is too complicated if you don't have any optional parameters for the decorator. Just use this kind of decorator which is way simpler:
https://github.com/DataDog/mortar-web/blob/40b5397656c70f9bedb42300d48993734d41cf53/mhc/lib/mongo.py#L72-L86

File: luigi/pubsub_history.py [hunk 3]
```diff
@@ -14,34 +14,95 @@
...
+
+        return wrapper
+
+    if function_to_decorate:
+        return _decorate(function_to_decorate)
+
+    return _decorate
+
+
+PUBSUB_SCOPES = ['https://www.googleapis.com/auth/pubsub']
```
Comment (antonipp) [inline luigi/pubsub_history.py]: Why is this a global variable if it's only used in once place?

File: luigi/pubsub_history.py [hunk 4]
```diff
@@ -14,34 +14,95 @@
...
     def _set_topic(self, project, topic):
         self.topic_path = 'projects/{}/topics/{}'.format(project, topic)
 
+    @retry_with_backoff()
     def _send_message(self, body):
-        return self.publisher.publish(self.topic_path, b64encode(json.dumps(body).encode('utf-8')))
+        pubsub_body = {'messages': [{'data': b64encode(str(body))}]}
+        response = self.client.projects().topics().publish(
+            topic=self.topic_path, body=pubsub_body).execute()
+        return "messageIds" in response
```
Comment (antonipp) [inline luigi/pubsub_history.py]: Are we doing something with these ids?

File: luigi/pubsub_history.py (hunk 2 above)
Comment (eplanet) [reply luigi/pubsub_history.py]: I don't think it's way simpler but I don't mind 😄

File: luigi/pubsub_history.py (hunk 4 above)
Comment (eplanet) [reply luigi/pubsub_history.py]: No.

Comment (antonipp) [reply luigi/pubsub_history.py]: No need to return them then 😄

File: luigi/pubsub_history.py [hunk 5]
```diff
@@ -16,32 +16,69 @@
...
+def retry_with_backoff(func):
+    """
+    Decorator to reconnect several times when an AutoReconnect exception occurs.
+    """
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        for sleep_time in TIME_TO_SLEEP:
+            try:
+                return func(*args, **kwargs)
+            except AutoReconnect, e:
```
Comment (antonipp) [inline luigi/pubsub_history.py]: This needs to be a normal Exception

File: luigi/pubsub_history.py [hunk 6]
```diff
@@ -16,32 +16,69 @@
...
+                if sleep_time == STOP_SLEEPING_INDICATOR:
+                    # raise the exception
+                    _logger.error('Failed to run mongo operation after several retries: %s' % e)
```
Comment (antonipp) [inline luigi/pubsub_history.py]: it's not mongo either 😄

File: luigi/pubsub_history.py [hunk 7]
```diff
@@ -16,32 +16,69 @@
...
 
 logger = logging.getLogger('luigi-interface')
 
+STOP_SLEEPING_INDICATOR = -1
+TIME_TO_SLEEP = [1.0, 2.0, 3.0, 5.0, 8.0, 13.0, STOP_SLEEPING_INDICATOR]
+
+
+def retry_with_backoff(func):
+    """
+    Decorator to reconnect several times when an AutoReconnect exception occurs.
```
Comment (antonipp) [inline luigi/pubsub_history.py]:
```suggestion
    Decorator to reconnect several times when an Exception occurs.
```

File: luigi/pubsub_history.py (hunk 4 above)
Comment (eplanet) [reply luigi/pubsub_history.py]: It doesn't return them, just `True` that there are `messageIds` in the response, meaning messages were actually sent.

Comment (antonipp) [reply luigi/pubsub_history.py]: oh right, I misread. but I meant we don't do anything with the return value of `_send_message`, so we might as well remove the `return` statement.

**Pull request merged**